// @sanity/default-layout/lib/navbar/search/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VIRTUAL_LIST_OVERSCAN = exports.VIRTUAL_LIST_ITEM_HEIGHT = exports.SEARCH_LIMIT = exports.POPOVER_MAX_WIDTH = exports.POPOVER_MAX_HEIGHT = exports.POPOVER_INPUT_PADDING = exports.GLOBAL_SEARCH_KEY_MODIFIER = exports.GLOBAL_SEARCH_KEY = exports.FINDABILITY_MVI = void 0;

var _ui = require("@sanity/ui");

var IS_MAC = typeof window != 'undefined' && /Mac|iPod|iPhone|iPad/.test(window.navigator.platform); // Findability version, prepended to every search query for future measurement

var FINDABILITY_MVI = 1; // Max number of records to fetch per search request - intentionally set high to workaround
// current limitations with studio search config

exports.FINDABILITY_MVI = FINDABILITY_MVI;
var SEARCH_LIMIT = 1000;
exports.SEARCH_LIMIT = SEARCH_LIMIT;
var GLOBAL_SEARCH_KEY = 'k';
exports.GLOBAL_SEARCH_KEY = GLOBAL_SEARCH_KEY;
var GLOBAL_SEARCH_KEY_MODIFIER = IS_MAC ? 'Cmd' : 'Ctrl';
exports.GLOBAL_SEARCH_KEY_MODIFIER = GLOBAL_SEARCH_KEY_MODIFIER;
var POPOVER_INPUT_PADDING = _ui.studioTheme.space[1]; // px

exports.POPOVER_INPUT_PADDING = POPOVER_INPUT_PADDING;
var POPOVER_MAX_HEIGHT = 735; // px

exports.POPOVER_MAX_HEIGHT = POPOVER_MAX_HEIGHT;
var POPOVER_MAX_WIDTH = 800; // px

exports.POPOVER_MAX_WIDTH = POPOVER_MAX_WIDTH;
var VIRTUAL_LIST_ITEM_HEIGHT = 55; // px

exports.VIRTUAL_LIST_ITEM_HEIGHT = VIRTUAL_LIST_ITEM_HEIGHT;
var VIRTUAL_LIST_OVERSCAN = 4;
exports.VIRTUAL_LIST_OVERSCAN = VIRTUAL_LIST_OVERSCAN;
